File: module_zhihu/src/main/java/com/gas/zhihu/ui/map/MapActivity.java
package com.gas.zhihu.ui.map;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.base.baseui.dialog.CommonDialog;
import com.base.baseui.utils.ViewUtils;
import com.base.lib.base.BaseActivity;
import com.base.lib.di.component.AppComponent;
import com.base.lib.util.ArmsUtils;
import com.gas.zhihu.R;
import com.gas.zhihu.R2;
import com.gas.zhihu.dialog.TipShowDialog;
import com.gas.zhihu.ui.map.di.DaggerMapComponent;
import com.gas.zhihu.ui.map.mvp.MapContract;
import com.gas.zhihu.ui.map.mvp.MapPresenter;
import com.gas.zhihu.ui.show.ShowActivity;
import com.lib.commonsdk.utils.GasAppUtils;
import com.tbruyelle.rxpermissions2.Permission;
import com.tbruyelle.rxpermissions2.RxPermissions;

import java.util.Timer;
import java.util.TimerTask;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

import static com.base.lib.util.Preconditions.checkNotNull;


/**
 * ================================================
 * desc: 地图搜索功能
 * <p>
 * created by author ljx
 * date  2020/3/28
 * email 569932357@qq.com
 * <p>
 * ================================================
 */

public class MapActivity extends BaseActivity<MapPresenter> implements MapContract.View {

    @BindView(R2.id.et_input)
    EditText etInput;
    private Context mContext;

    @Inject
    RxPermissions mRxPermissions;

    @Override
    public void setupActivityComponent(@NonNull AppComponent appComponent) {
        DaggerMapComponent //如找不到该类,请编译一下项目
                .builder()
                .appComponent(appComponent)
                .view(this)
                .build()
                .inject(this);
    }

    @Override
    public int initView(@Nullable Bundle savedInstanceState) {
        //如果你不需要框架帮你设置 setContentView(id) 需要自行设置,请返回 0
        mContext = this;
        return R.layout.zhihu_activity_map;

    }

    @Override
    public void initData(@Nullable Bundle savedInstanceState) {

    }


    @Override
    public void showMessage(@NonNull String message) {
        checkNotNull(message);
        ArmsUtils.snackbarText(message);
    }

    @Override
    public void launchActivity(@NonNull Intent intent) {
        checkNotNull(intent);
        ArmsUtils.startActivity(intent);
    }

    @Override
    public void killMyself() {
        finish();
    }



    @OnClick({R2.id.iv_clear, R2.id.btn_go_login, R2.id.text_see_match_rule})
    public void onViewClicked(View view) {
        if (view.getId() == R.id.iv_clear) {
            etInput.setText("");
        } else if (view.getId() == R.id.btn_go_login) {
            toSearch();

        } else if (view.getId() == R.id.text_see_match_rule) {
            showRuleDialog();

        }
    }




    private void showRuleDialog() {
        new TipShowDialog().show(this,"提示","匹配规则的提示");

    }

    private void toSearch() {

        String search = ViewUtils.getNoSpaceText(etInput);

        if (TextUtils.isEmpty(search)) {
            GasAppUtils.toast("请输入正确的匹配规则");
            return;
        }
        ShowActivity.Companion.launchActivity(mContext,search);

    }

    /**
     * 双击退出函数
     */
    private static Boolean isExit = false;

    private void exitBy2Click() {
        Timer tExit;
        if (!isExit) {
            isExit = true; // 准备退出
            ArmsUtils.snackbarText("再按一次退出程序");
            tExit = new Timer();
            tExit.schedule(new TimerTask() {
                @Override
                public void run() {
                    isExit = false; // 取消退出
                }
            }, 2000); //如果2秒钟内没有按下返回键，则启动定时器取消掉刚才执行的任务

        } else {
            ArmsUtils.exitApp();
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            exitBy2Click();      //调用双击退出函数
        }
        return false;
    }




    @Override
    public RxPermissions getRxPermissions() {
        return mRxPermissions;
    }

    @Override
    public Activity getActivity() {
        return this;
    }
}

File: module_zhihu/src/main/java/com/gas/zhihu/ui/map/MapActivity.kt
package com.gas.zhihu.ui.map

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import com.base.baseui.dialog.CommonDialog
import com.base.baseui.dialog.QMUITipDialog
import com.base.baseui.utils.ViewUtils
import com.base.lib.base.BaseActivity
import com.base.lib.di.component.AppComponent
import com.base.lib.util.ArmsUtils
import com.base.lib.util.Preconditions
import com.gas.zhihu.R
import com.gas.zhihu.dialog.TextFilterDialog
import com.gas.zhihu.ui.detial.DetailsActivity
import com.gas.zhihu.ui.map.di.DaggerMapComponent
import com.gas.zhihu.ui.map.mvp.MapContract
import com.gas.zhihu.ui.map.mvp.MapPresenter
import com.gas.zhihu.ui.show.ShowActivity.Companion.launchActivity
import com.google.android.flexbox.FlexboxLayoutManager
import com.lib.commonsdk.utils.AppUtils
import com.tbruyelle.rxpermissions2.RxPermissions
import io.reactivex.ObservableOnSubscribe
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.zhihu_activity_map.*
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

/**
 * ================================================
 * desc: 地图搜索功能
 *
 *
 * created by author ljx
 * date  2020/3/28
 * email 569932357@qq.com
 *
 *
 * ================================================
 */
class MapActivity : BaseActivity<MapPresenter?>(), MapContract.View {


    private var mContext: Context? = null

    @Inject
    lateinit var mRxPermissions: RxPermissions
    @Inject
    lateinit var mManager: FlexboxLayoutManager
    @Inject
    lateinit var mAdapter: SearchRecordAdapter

    private val loadDialog: QMUITipDialog by lazy {
        QMUITipDialog.Builder(mContext)
                .setIconType(QMUITipDialog.Builder.ICON_TYPE_LOADING)
                .setTipWord("正在加载")
                .create()
    }

    private lateinit var mPublishSubject: PublishSubject<String>
    private val mCompositeDisposable by lazy {
        CompositeDisposable()
    }

    private val popupText: TextFilterDialog by lazy {
        TextFilterDialog()
    }


    override fun setupActivityComponent(appComponent: AppComponent) {
        DaggerMapComponent //如找不到该类,请编译一下项目
                .builder()
                .appComponent(appComponent)
                .view(this)
                .build()
                .inject(this)
    }

    override fun initView(savedInstanceState: Bundle?): Int { //如果你不需要框架帮你设置 setContentView(id) 需要自行设置,请返回 0
        mContext = this
        return R.layout.zhihu_activity_map
    }

    override fun initData(savedInstanceState: Bundle?) {
        viewInit()
    }

    private fun viewInit() {

        title_view.setOnRightListener {

        }
        imgDeleteRecord.setOnClickListener {
            showDeleteTipDialog()
        }
        btn_go_login.setOnClickListener {
            val search = ViewUtils.getNoSpaceText(et_input)
            toSearch(search, true)
        }
        text_see_match_rule.setOnClickListener {
            startActivity(Intent(this, DetailsActivity::class.java))
        }
        mAdapter.setOnMultiItemClickListener { _, data, position, _ ->
            mAdapter.onRecordItemClick(position)
            mPresenter!!.searchRecord = mAdapter.data
            toSearch(data!!, false)
        }
        rvSearchHistory.layoutManager = mManager
        rvSearchHistory.adapter = mAdapter
        val list = mPresenter!!.searchRecord
        if (list == null || list.isEmpty()) {
            vRecord.visibility = View.GONE
        } else {
            mAdapter.setInitData(list)
            vRecord.visibility = View.VISIBLE
        }
        et_input.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                if (s.toString().trim().isEmpty()) {
                    startSearch("")
                } else {
                    //输入内容非空的时候才开始搜索
                    startSearch(s.toString())
                }
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

        })


        mPublishSubject = PublishSubject.create()
        mCompositeDisposable.add(mPublishSubject.debounce(200, TimeUnit.MILLISECONDS)
//                .filter { t -> t.isNotEmpty() }
                .switchMap { t -> getSearchObservable(t) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({showSearchResult(it)},{
                    it.printStackTrace()
                }))

        popupText.setListener {
            et_input.setText(it)
            et_input.setSelection(it.length)
            startSearch("")
            popupText.dismiss()
        }
    }

    private fun getSearchObservable(query: String):Observable<List<String>> {

        if(query.isEmpty()){
            return Observable.empty()
        }
       return Observable.create(ObservableOnSubscribe<List<String>> { emitter ->
           emitter.onNext(mPresenter!!.getSearchKeysWithKey(query))
           emitter.onComplete()
       }).subscribeOn(Schedulers.io())
    }
    private fun startSearch(str: String) {
        mPublishSubject.onNext(str)
    }

    private fun showSearchResult(list: List<String>) {
        popupText.apply {
            if (isShow()) {
                updateData(list)
            } else {
                show(et_input, list)
            }
        }
    }

    override fun showMessage(message: String) {
        Preconditions.checkNotNull(message)
        ArmsUtils.snackbarText(message)
    }

    override fun launchActivity(intent: Intent) {
        Preconditions.checkNotNull(intent)
        ArmsUtils.startActivity(intent)
    }

    override fun killMyself() {
        finish()
    }

    private fun showDeleteTipDialog() {
        CommonDialog.Builder(mContext)
                .setTitle("是否删除搜索记录")
                .setLeftTitle("取消")
                .setRightTitle("确定")
                .setCancelable(true)
                .setDialogClickListener(object : CommonDialog.onDialogClickListener {
                    override fun onRightClick() {
                        vRecord.visibility = View.GONE
                        mPresenter?.clearSearchRecord()
                        mAdapter.clearData()
                    }
                }).create().apply {
                    show()
                }

    }

    private fun toSearch(search: String, needToSave: Boolean) {
        if (TextUtils.isEmpty(search)) {
            AppUtils.toast("请输入正确的匹配规则")
            return
        }
        if (needToSave) {
            if (vRecord!!.visibility == View.GONE) {
                vRecord!!.visibility = View.VISIBLE
            }
            mAdapter!!.addItem(search)
            mPresenter!!.searchRecord = mAdapter!!.data
        }
        launchActivity(mContext, search)
    }





    override fun showLoading() {
        loadDialog.show()
    }

    override fun hideLoading() {
        loadDialog.dismiss()
    }

    override fun getRxPermissions(): RxPermissions {
        return mRxPermissions!!
    }

    override fun getActivity(): Activity {
        return this
    }


    companion object {
        /**
         * 双击退出函数
         */
        private var isExit = false
    }
}